thrifty-kotlin-codegen/src/main/kotlin/com/microsoft/thrifty/kgen/KotlinCodeGenerator.kt

```diff
@@ -1974,6 +2002,174 @@ class KotlinCodeGenerator(
     }
 
+    internal fun generateProcessorImplementation(schema: Schema, serviceType: ServiceType, serviceInterface: TypeSpec, specsByNamespace: LinkedHashMultimap<String, TypeSpec>): TypeSpec {
+        val serverTypeName = serviceType.name + "Processor"
+        val type = TypeSpec.classBuilder(serverTypeName).apply {
+            primaryConstructor(
+                    FunSpec.constructorBuilder()
+                            .addParameter("handler", serviceType.typeName)
+                            .addParameter(ParameterSpec.builder(
+                                    "errorHandler",
+                                    ErrorHandler::class,
+                                    KModifier.PRIVATE).defaultValue("%T", DefaultErrorHandler::class).build()
+                            )
+                            .build()
+            )
+            addProperty(
+                    PropertySpec.builder("errorHandler", ErrorHandler::class)
+                            .initializer("errorHandler")
+                            .addModifiers(KModifier.PRIVATE)
+                            .build()
+            )
+            addProperty(
+                    PropertySpec.builder("handler", serviceType.typeName)
+                            .initializer("handler")
+                            .addModifiers(KModifier.PRIVATE)
+                            .build()
+            )
+            addSuperinterface(Processor::class)
+        }
+
+        val spec = FunSpec.builder("process").apply {
+            addModifiers(KModifier.SUSPEND, KModifier.OVERRIDE)
+            addParameter("input", Protocol::class)
+            addParameter("output", Protocol::class)
+        }
+
+        spec.addCode {
+            beginControlFlow("input.%M { msg ->", MemberName("com.microsoft.thrifty.service.server", "readMessage"))
+            beginControlFlow("val call = when(msg.name)")
+        }
+
+        serviceInterface.funSpecs.zip(serviceType.methods).forEach { (interfaceFun, method) ->
+            val argsDataClass = generateDataClass(schema, method.argsStruct)
+            val resultDataClass = if (method.resultStruct.fields.isEmpty()) generateDataClass(schema, method.resultStruct) else generateSealedClass(schema, method.resultStruct)
+            specsByNamespace.put(serviceType.kotlinNamespace, argsDataClass)
+            specsByNamespace.put(serviceType.kotlinNamespace, resultDataClass)
+
+            val call = buildServerCallType(schema, method, interfaceFun, argsDataClass, resultDataClass)
+            type.addType(call)
+
+            spec.addCode {
+                addStatement( "%S -> %T()",
+                        method.name,
+                        ClassName(serviceType.kotlinNamespace, serverTypeName, callTypeName(method)))
+            }
+        }
+
+        spec.addCode {
+            beginControlFlow("else -> ")
+            addStatement("""throw %T("%L")""", IllegalArgumentException::class, "Unknown method \${msg.name}")
```

Comment:
Be careful with using `Foo::class` as type arguments.  Kotlinpoet has an iffy track record with understanding, for example `kotlin.Exception` - historically it will often end up emitting references to the underlying JVM type, which is _ok_ I guess for this use-case but forecloses on MPP servers too.

tl;dr let's add an entry to the `ClassNames` object up around line 110 of this file and use that instead.

Reply:
I see a few other places in the PR (and probably in my code, too, let's be honest) where `::class` literals are being used this way.  So long as they aren't typealiases or expect/actual types, it's probably OK, but we should make sure.

Reply:
Honestly I have almost no experience with Kotlin Multiplatform, so I tried to do what you said to the places I found :)